ui/src/vayla-design-lib/datepicker/datepicker.tsx

```diff
@@ -6,13 +6,78 @@ import { fi } from 'date-fns/locale';
 import { IconColor, Icons, IconSize } from 'vayla-design-lib/icon/Icon';
 import { createClassName } from 'vayla-design-lib/utils';
 import { TextField, TextInputIconPosition } from 'vayla-design-lib/text-field/text-field';
+import { CloseableModal } from 'vayla-design-lib/closeable-modal/closeable-modal';
+import { formatDateShort } from 'utils/date-utils';
+import { parse, isValid } from 'date-fns';
 import { useCloneRef } from 'utils/react-utils';
 
 type DatePickerProps = {
     value: Date | undefined;
-    onChange?: (date: Date | undefined) => void;
+    onChange: (date: Date | undefined) => void;
+    wide?: boolean;
 } & Omit<React.InputHTMLAttributes<HTMLInputElement>, 'value' | 'onChange'>;
 
+type DatePickerInputProps = {
+    openDatePicker: () => void;
+    date: Date | undefined;
+    setDate: (date: Date | undefined) => void;
+    wide: boolean | undefined;
+} & Omit<React.InputHTMLAttributes<HTMLInputElement>, 'value' | 'onChange'>;
+
+const DatePickerInput = React.forwardRef<HTMLInputElement, DatePickerInputProps>(
+    ({ openDatePicker, date, setDate, wide, ...props }, ref) => {
+        const [value, setValue] = React.useState<string>('');
+        const localRef = useCloneRef<HTMLInputElement>(ref);
+        React.useEffect(() => {
+            if (document.activeElement !== localRef.current) {
+                setValue(date ? formatDateShort(date) : '');
+            }
+        }, [date]);
+
+        function setValueAndSetDateIfValid(e: React.ChangeEvent<HTMLInputElement>): void {
+            setValue(e.target.value);
+
+            const newDate = parse(e.target.value, 'dd.MM.yyyy', new Date());
+            if (isValid(newDate)) {
+                setDate(newDate);
+            }
+        }
+
+        function setDateOrResetIfInvalid(e: React.FocusEvent<HTMLInputElement>): void {
+            const newDate = parse(e.target.value, 'dd.MM.yyyy', new Date());
+            if (isValid(newDate)) {
+                setDate(newDate);
+            } else {
+                setValue(date ? formatDateShort(date) : '');
+            }
+        }
```

Comment:
Tekstikenttä toimii siis näin (aiempi toiminnallisuus pyritty säilyttämään):
- Tekstikenttään voi kirjoittaa mitä vaan, vaikka "blaa"
- Aina kun tekstikentässä on validi päivämäärä, ka. päivämäärä asetetaan valituksi päivämääräksi vaikka kyseessä olisi vain välitulos
- Kun tekstikentästä lähdetään pois, jos päivämääräkentässä ei ole oikeaa arvoa, tekstikentän sisältö resetataan viimeisimpään valittuun päivämäärään (mikäli sellainen on olemassa, muulloin resetataan tyhjään)